git_machete/cli.py

```diff
@@ -75,274 +77,423 @@ def usage(command: str = None) -> None:
+            subparser_name = full_parser_name_splited[1]
+            usage(subparser_name)
+        else:
+            usage()
+        parser.exit()
+
+
+def create_cli_parser() -> argparse.ArgumentParser:
+    cli_parser = argparse.ArgumentParser(
+        prog='git-machete', argument_default=argparse.SUPPRESS, add_help=False)
+    cli_parser.add_argument('--debug', action='store_true')
+    cli_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+    cli_parser.add_argument(
+        '--version', action='version', version=f'%(prog)s version {__version__}')
+    cli_parser.add_argument('-v', '--verbose', action='store_true')
+
+    subparsers = cli_parser.add_subparsers(dest='command')
+
+    add_parser = subparsers.add_parser('add', argument_default=argparse.SUPPRESS, add_help=False)
+    add_parser.add_argument('branch', nargs='?')
+    add_parser.add_argument('-o', '--onto')
+    add_parser.add_argument('-R', '--as-root', action='store_true')
+    add_parser.add_argument('-y', '--yes', action='store_true')
+    add_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
+    advance_parser = subparsers.add_parser('advance', add_help=False)
+    advance_parser.add_argument('-y', '--yes', action='store_true', default=argparse.SUPPRESS)
+    advance_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
+    anno_parser = subparsers.add_parser('anno', add_help=False)
+    # possible values of 'annotation_text' include: [], [''], ['some_val'], ['text_1', 'text_2']
+    anno_parser.add_argument('annotation_text', nargs='*')
+    anno_parser.add_argument('-b', '--branch', default=argparse.SUPPRESS)
+    anno_parser.add_argument(
+        '-H', '--sync-github-prs', action='store_true', default=argparse.SUPPRESS)
+    anno_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
+    delete_unmanaged_parser = subparsers.add_parser('delete-unmanaged', add_help=False)
+    delete_unmanaged_parser.add_argument(
+        '-y', '--yes', action='store_true', default=argparse.SUPPRESS)
+    delete_unmanaged_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
+    diff_full_parser = subparsers.add_parser('diff', aliases=['d'], add_help=False)
+    diff_full_parser.add_argument('branch', nargs='?')
+    diff_full_parser.add_argument(
+        '-s', '--stat', action='store_true', default=argparse.SUPPRESS)
+    diff_full_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
+    discover_parser = subparsers.add_parser(
+        'discover', argument_default=argparse.SUPPRESS, add_help=False)
+    discover_parser.add_argument('-C', '--check-out-since')
+    discover_parser.add_argument('-l', '--list-commits', action='store_true')
+    discover_parser.add_argument('-r', '--roots')
+    discover_parser.add_argument('-y', '--yes', action='store_true')
+    discover_parser.add_argument('-h', '--help', action=MacheteHelpAction)
+
```

Comment:
Not-so-nit: same goes for `-v` and other common flags, as for now they can't be passed at the end:

```
$ git machete github anno-prs --debug
usage: git-machete [--debug] [-h] [--version] [-v]
                   {add,advance,anno,delete-unmanaged,diff,d,discover,edit,e,file,fork-point,format,github,go,g,help,hooks,is-managed,list,log,l,reapply,show,slide-out,squash,status,s,traverse,update,version}
                   ...
git-machete: error: unrecognized arguments: --debug
``

Reply:
We may have a problem, because this is not be possible -> https://docs.python.org/3/library/argparse.html#intermixed-parsing 
I'm using subparsers and mutually exclusive groups. Mutually exclusive groups could be probably dropped and replaced with something primitive, but dropping subparsers invalidates the idea of using argparse.

Reply:
Hmmm okay but even without this intermixed parsing... how about just adding `<subcommand>_parser.add_argument('-v', '--verbose', action='store_true')` to every single subparser (probably in some loop), similarly as it's done for `--help`?

Reply:
I found a counterintuitive hack. So there will be now a new parser, with just these common optional args like -v -h --debug that will be kind of joined with all other parsers (kind of like inheritance). It will work like that:

```
➜  git-machete git:(improve/parser) git machete --debug show up
>>> git symbolic-ref --quiet HEAD
<stdout>:
refs/heads/improve/parser

enchance/checkout-pr
➜  git-machete git:(improve/parser) git machete show up --debug
>>> git symbolic-ref --quiet HEAD
<stdout>:
refs/heads/improve/parser

enchance/checkout-pr
➜  git-machete git:(improve/parser) git machete show --debug up        
>>> git symbolic-ref --quiet HEAD
<stdout>:
refs/heads/improve/parser
```

Reply:
> Hmmm okay but even without this intermixed parsing... how about just adding `<subcommand>_parser.add_argument('-v', '--verbose', action='store_true')` to every single subparser (probably in some loop), similarly as it's done for `--help`?

Sorry, my gh was not refreshed and I missed that comment. Responding: it's no longer needed. This intermixed stuff is confusing, but seems to be not a problem, when there is a separate generic_optional_args_parser, so no need to add help or version to each subparser separately.